rustworkx-core/src/dag_algo.rs

```diff
@@ -314,6 +316,140 @@ where
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LayersIter<G, N> {
+    graph: G,
+    cur_layer: Vec<N>,
+    next_layer: Vec<N>,
+    predecessor_count: HashMap<N, usize>,
+    first_iter: bool,
+    cycle_check: HashSet<Vec<N>>,
+}
+
+impl<G, N> Iterator for LayersIter<G, N>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected // Used for .edged_directed
+        + GraphBase<NodeId = N>,
+    N: Debug + Copy + Eq + Hash,
+{
+    type Item = Vec<N>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.first_iter {
+            self.first_iter = false;
+            self.cycle_check.insert(self.cur_layer.clone());
+            Some(self.cur_layer.to_owned())
+        } else if self.cur_layer.is_empty() {
+            None
+        } else {
+            for node in &self.cur_layer {
+                if self.graph.to_index(*node) >= self.graph.node_bound() {
+                    panic!("Node {:#?} is not present in the graph.", node);
+                }
+                let children = self
+                    .graph
+                    .neighbors_directed(*node, petgraph::Direction::Outgoing);
+                let mut used_indices: HashSet<G::NodeId> = HashSet::new();
+                for succ in children {
+                    // Skip duplicate successors
+                    if used_indices.contains(&succ) {
+                        continue;
+                    }
+                    used_indices.insert(succ);
+                    let mut multiplicity: usize = 0;
+                    let raw_edges: G::EdgesDirected = self
+                        .graph
+                        .edges_directed(*node, petgraph::Direction::Outgoing);
+                    for edge in raw_edges {
+                        if edge.target() == succ {
+                            multiplicity += 1;
+                        }
+                    }
+                    self.predecessor_count
+                        .entry(succ)
+                        .and_modify(|e| *e -= multiplicity)
+                        .or_insert(
+                            // Get the number of incoming edges to the successor
+                            self.graph
+                                .edges_directed(succ, petgraph::Direction::Incoming)
+                                .count()
+                                - multiplicity,
+                        );
+                    if *self.predecessor_count.get(&succ).unwrap() == 0 {
+                        self.next_layer.push(succ);
+                        self.predecessor_count.remove(&succ);
+                    }
+                }
+            }
+            swap(&mut self.cur_layer, &mut self.next_layer);
+            if self.cycle_check.contains(&self.cur_layer) {
+                panic!("The layers algorithm doesn't support graphs with cycles.")
```

Comment:
I'm starting to think this should be a result now actually; I apologize this was my earlier mistake. I was thinking about how we document this on the python side, and the answer is we need to raise an exception to python, and that's tricky to do if we panic. I just really don't like iterators with `Result<T>` items. But at the end of the day doing:
```
let layers: Result<Vec<_>> = layers().map().collect();
```
isn't terrible I guess.

Reply:
Sounds good, I will continue to evaluate this in that case.